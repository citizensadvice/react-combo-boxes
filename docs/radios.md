# `<Radios>`

Generates a group of radios.

Although radios are not a combo-box they are a good alternative.
Many users find combo-boxes hard to use and radios perform a very
similar function to a `<select>`.

## Usage

This is a stateless controlled component.  You must respond to `onValue` to update the selected value.

```js
const [value, setValue] = useState(initialValue);

<Radios
  options={options}
  value={value}
  onValue={setValue}
/>
```

## Props

### `name: String`

`name` attribute for each radio. This is required by HTML for a radio group to function.

### `options: Array<any>`

The options to display. [See options][options].  

Radios take an additional `description` property that allows an optional description to be added to a radio.

### `mapOption: (option: any) => Option | string`

Map an option into label, value, disabled and group. [See options][options]  

### `value: any`

The currently selected option

### `onChange: (event: Event) => void`

Change event added to each radio.

### `onValue: (option: any) => void`

Called with the selected option when a user selects an option.

#### `required: boolean = false`

Mark all the radios in the radio group as required.

#### `classPrefix: string = "react-combo-boxes-radio'`

By default the component is built using BEM style class names.  This sets the prefix for all those classes.

Set to falsey to remove the BEM classes.

#### `groupClassPrefix: string = "react-combo-boxes-radio-group'`

BEM class prefix for groups.

Set to falsey to remove the BEM classes.

#### render props

A number of render methods are provided to customise the appearance of the component.

Each render method has the signature

```js
(
  // The props set on the <ComboBox>
  props: object,
  // The state of the component
  state: {
    // Current option
    option: Option;
    // Is the current option selected
    checked: boolean;
    // Current group
    group: Group;
  },
  // The props passed into the component
  componentProps: object;
) => React.ReactNode
```

The render functions available are:

| Name                         | Default value                     | Description                                                                         |
| ----                         | ----                              | ----                                                                                |
| `renderWrapper`              | `(props) => <div {...props} />`   | Renders the wrapper for each radio                                                  |
| `renderInput`                | `(props) => <input {...props} />` | Renders the radio input                                                             |
| `renderLabel`                | `(props) => <label {...props} />` | Renders label for a radio                                                           |
| `renderDescription`          | `(props) => <div {...props} />`   | Renders the optional description for a radio                                        |
| `renderGroup`                | `(props) => <div {...props} />`   | Wraps a group of options                                                            |
| `renderGroupLabel`           | `(props) => <div {...props} />`   | Renders the visible label for a group. This will be ignored by a screen-reader      |
| `renderGroupAccessibleLabel` | `(props) => <span {...props} />`  | Renders the accessible label for a group.  This will be read out before each option |


[options]: options.md
