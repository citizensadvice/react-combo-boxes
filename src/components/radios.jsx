import { Fragment, useCallback, memo } from 'react';
import PropTypes from 'prop-types';
import { useNormalisedOptions } from '../hooks/use_normalised_options';
import { renderGroupedOptions } from '../helpers/render_grouped_options';
import { classPrefix as defaultClassPrefix } from '../constants/class_prefix';
import { makeBEMClass } from '../helpers/make_bem_class';
import { visuallyHiddenClassName } from '../constants/visually_hidden_class_name';

const defaultRenderWrapper = ({ key, ...props }) => (
  <div
    key={key}
    {...props}
  />
);
const defaultRenderInput = (props) => <input {...props} />;
const defaultRenderLabel = (props) => <label {...props} />;
// eslint-disable-next-line react/jsx-no-useless-fragment
const defaultRenderLabelWrapper = (props) => <Fragment {...props} />;
const defaultRenderDescription = (props) => <div {...props} />;
const defaultRenderGroup = ({ key, ...props }) => (
  <div
    key={key}
    {...props}
  />
);
const defaultRenderGroupLabel = (props) => <div {...props} />;
const defaultRenderGroupAccessibleLabel = (props) => <span {...props} />;

export const Radios = memo(
  ({
    classPrefix = `${defaultClassPrefix}radio`,
    groupClassPrefix = `${defaultClassPrefix}radio-group`,
    onChange = null,
    onValue = null,
    name,
    renderWrapper = defaultRenderWrapper,
    renderInput = defaultRenderInput,
    renderLabel = defaultRenderLabel,
    renderLabelWrapper = defaultRenderLabelWrapper,
    renderDescription = defaultRenderDescription,
    renderGroup = defaultRenderGroup,
    renderGroupLabel = defaultRenderGroupLabel,
    renderGroupAccessibleLabel = defaultRenderGroupAccessibleLabel,
    required = false,
    value: _value = null,
    ...rawProps
  }) => {
    const optionisedProps = Object.freeze({
      classPrefix,
      groupClassPrefix,
      onChange,
      onValue,
      renderWrapper,
      renderInput,
      renderLabel,
      renderLabelWrapper,
      renderDescription,
      renderGroup,
      renderGroupLabel,
      renderGroupAccessibleLabel,
      required,
      ...rawProps,
      ...useNormalisedOptions({
        ...rawProps,
        value: _value,
        placeholderOption: undefined,
      }),
    });

    const { options, selectedOption } = optionisedProps;

    const handleChange = useCallback(
      (e) => {
        onValue?.(options.find((o) => o.identity === e.target.value)?.value);
        onChange?.(e);
      },
      [onChange, onValue, options],
    );

    return renderGroupedOptions({
      options,
      renderGroup(group) {
        const { children, key, label, html } = group;
        return renderGroup(
          {
            key,
            className: makeBEMClass(groupClassPrefix),
            children: (
              <>
                {renderGroupLabel(
                  {
                    children: label,
                    className: makeBEMClass(groupClassPrefix, 'label'),
                    ...html,
                  },
                  { group },
                  optionisedProps,
                )}
                {children}
              </>
            ),
          },
          { group },
          optionisedProps,
        );
      },
      renderOption(option) {
        const { identity, label, key, html, disabled, description, group } =
          option;
        const checked = selectedOption?.identity === identity;

        return renderWrapper(
          {
            className: makeBEMClass(classPrefix),
            children: (
              <>
                {renderInput(
                  {
                    'aria-describedby': description
                      ? `${key}_description`
                      : null,
                    checked,
                    disabled,
                    id: key,
                    name,
                    onChange: handleChange,
                    type: 'radio',
                    value: identity,
                    required,
                    className: makeBEMClass(classPrefix, 'input'),
                    ...html,
                  },
                  { option, checked },
                  optionisedProps,
                )}
                {renderLabelWrapper({
                  children: (
                    <>
                      {renderLabel(
                        {
                          htmlFor: key,
                          children: (
                            <>
                              {group
                                ? renderGroupAccessibleLabel(
                                    {
                                      className: visuallyHiddenClassName,
                                      children: `${group.label} `,
                                    },
                                    { group },
                                    optionisedProps,
                                  )
                                : null}
                              {label}
                            </>
                          ),
                          className: makeBEMClass(classPrefix, 'label'),
                        },
                        { option, checked },
                        optionisedProps,
                      )}
                      {description && ' '}
                      {!!description &&
                        renderDescription(
                          {
                            id: `${key}_description`,
                            children: description,
                            className: makeBEMClass(classPrefix, 'description'),
                          },
                          { option, group, checked },
                          optionisedProps,
                        )}
                    </>
                  ),
                })}
              </>
            ),
            key,
          },
          { option, group, checked },
          optionisedProps,
        );
      },
    });
  },
);

Radios.propTypes = {
  classPrefix: PropTypes.string,
  groupClassPrefix: PropTypes.string,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onValue: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.any).isRequired,
  value: PropTypes.any,
  required: PropTypes.bool,
  renderWrapper: PropTypes.func,
  renderInput: PropTypes.func,
  renderLabel: PropTypes.func,
  renderLabelWrapper: PropTypes.func,
  renderDescription: PropTypes.func,
  renderGroup: PropTypes.func,
  renderGroupLabel: PropTypes.func,
  renderGroupAccessibleLabel: PropTypes.func,
};

Radios.displayName = 'Radios';
