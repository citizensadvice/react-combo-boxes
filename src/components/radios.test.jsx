import { useState } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Radios } from './radios';

function Test({ value: _value, ...props }) {
  const [value, setValue] = useState(_value);

  return (
    <Radios
      id="id"
      name="name"
      value={value}
      onValue={setValue}
      {...props}
    />
  );
}

describe('options', () => {
  describe('as array of strings', () => {
    const options = ['Apple', 'Banana', 'Orange'];

    it('renders radio group', () => {
      const { container } = render(<Test options={options} />);
      expect(container).toMatchSnapshot();
    });

    it('renders radios with a selected value', () => {
      render(
        <Test
          options={options}
          value="Orange"
        />,
      );
      expect(screen.getByRole('radio', { checked: true })).toEqual(
        screen.getByRole('radio', { name: 'Orange' }),
      );
    });

    it('triggers the onValue callback with the selected value', async () => {
      const spy = jest.fn();
      render(
        <Test
          options={options}
          onValue={spy}
        />,
      );
      await userEvent.click(screen.getByRole('radio', { name: 'Banana' }));
      expect(spy).toHaveBeenCalledWith('Banana');
    });

    it('updates when the value changes', async () => {
      render(<Test options={options} />);
      const option = screen.getByRole('radio', { name: 'Banana' });
      await userEvent.click(option);
      expect(screen.getByRole('radio', { checked: true })).toEqual(option);
    });
  });

  describe('options as array of numbers', () => {
    const options = [0, 1, 2, 3];

    it('renders radio group', () => {
      const { container } = render(<Test options={options} />);
      expect(container).toMatchSnapshot();
    });

    it('renders radios with a selected value', () => {
      render(
        <Test
          options={options}
          value={2}
        />,
      );
      expect(screen.getByRole('radio', { checked: true })).toEqual(
        screen.getByRole('radio', { name: '2' }),
      );
    });

    it('triggers the onValue callback with the selected value', async () => {
      const spy = jest.fn();
      render(
        <Test
          options={options}
          onValue={spy}
        />,
      );
      await userEvent.click(screen.getByRole('radio', { name: '2' }));
      expect(spy).toHaveBeenCalledWith(2);
    });

    it('updates when the value changes', async () => {
      render(<Test options={options} />);
      const option = screen.getByRole('radio', { name: '2' });
      await userEvent.click(option);
      expect(screen.getByRole('radio', { checked: true })).toEqual(option);
    });
  });

  describe('options as array of objects', () => {
    describe('label', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Banana' },
        { label: 'Orange' },
      ];

      it('renders radio group', () => {
        const { container } = render(<Test options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('renders radios with a selected value', () => {
        render(
          <Test
            options={options}
            value={{ label: 'Banana' }}
          />,
        );
        expect(screen.getByRole('radio', { checked: true })).toEqual(
          screen.getByRole('radio', { name: 'Banana' }),
        );
      });

      it('triggers the onValue callback with the selected value', async () => {
        const spy = jest.fn();
        render(
          <Test
            options={options}
            onValue={spy}
          />,
        );
        await userEvent.click(screen.getByRole('radio', { name: 'Banana' }));
        expect(spy).toHaveBeenCalledWith({ label: 'Banana' });
      });

      it('updates when the value changes', async () => {
        render(<Test options={options} />);
        const option = screen.getByRole('radio', { name: 'Banana' });
        await userEvent.click(option);
        expect(screen.getByRole('radio', { checked: true })).toEqual(option);
      });
    });

    describe('value', () => {
      it('is used as a options identity', async () => {
        const options = [
          { label: 'foo', value: 1 },
          { label: 'foo', value: 2 },
          { label: 'foo', value: 3 },
        ];
        render(
          <Test
            options={options}
            value={2}
          />,
        );
        expect(screen.getByRole('radio', { checked: true }).value).toEqual('2');
        const radios = screen.getAllByRole('radio');
        await userEvent.click(radios[0]);
        expect(screen.getByRole('radio', { checked: true })).toEqual(radios[0]);
      });
    });

    describe('id', () => {
      it('is used as a options identity', async () => {
        const options = [
          { label: 'foo', id: 1 },
          { label: 'foo', id: 2 },
          { label: 'foo', id: 3 },
        ];
        render(
          <Test
            options={options}
            value={2}
          />,
        );
        expect(screen.getByRole('radio', { checked: true }).value).toEqual('2');
        const radios = screen.getAllByRole('radio');
        await userEvent.click(radios[0]);
        expect(screen.getByRole('radio', { checked: true })).toEqual(radios[0]);
      });
    });

    describe('disabled', () => {
      it('sets the disabled attribute', () => {
        const options = [{ label: 'foo', disabled: true }];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).toBeDisabled();
      });

      it('will display a disabled radio as checked', () => {
        const options = [{ label: 'foo', disabled: true }];
        render(
          <Test
            options={options}
            value="foo"
          />,
        );
        expect(screen.getByRole('radio')).toBeChecked();
      });
    });

    describe('description', () => {
      it('sets the description', () => {
        const options = [{ label: 'foo', description: 'foo bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).toHaveAccessibleDescription(
          'foo bar',
        );
      });
    });

    describe('hint', () => {
      it('is used as a fallback for description', () => {
        const options = [{ label: 'foo', hint: 'foo bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).toHaveAccessibleDescription(
          'foo bar',
        );
      });

      it('is not used if description is present', () => {
        const options = [
          { label: 'foo', description: 'fizz buzz', hint: 'foo bar' },
        ];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).toHaveAccessibleDescription(
          'fizz buzz',
        );
      });
    });

    describe('html', () => {
      it('sets attributes on the radio', () => {
        const options = [
          { label: 'foo', html: { 'data-foo': 'bar', className: 'class' } },
        ];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).toHaveAttribute('data-foo', 'bar');
        expect(screen.getByRole('radio')).toHaveClass('class');
      });
    });

    describe('group', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Orange', group: 'Citrus' },
        { label: 'Lemon', group: 'Citrus' },
        { label: 'Raspberry', group: 'Berry' },
        { label: 'Strawberry', group: 'Berry' },
      ];

      it('renders grouped options', () => {
        const { container } = render(<Test options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('selects an option', async () => {
        render(<Test options={options} />);
        const option = screen.getByRole('radio', { name: 'Citrus Lemon' });
        await userEvent.click(option);
        expect(screen.getByRole('radio', { checked: true })).toEqual(option);
      });
    });

    describe('other attributes', () => {
      it('does not render them', () => {
        const options = [{ label: 'foo', 'data-foo': 'bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('radio')).not.toHaveAttribute(
          'data-foo',
          'bar',
        );
      });
    });
  });

  describe('mapOption', () => {
    describe('mapOption returns an object', () => {
      const options = [
        { name: 'Apple' },
        { name: 'Banana' },
        { name: 'Orange' },
      ];

      it('maps options', async () => {
        render(
          <Test
            options={options}
            mapOption={({ name }) => ({ label: name })}
          />,
        );
        const option = screen.getByRole('radio', { name: 'Banana' });
        await userEvent.click(option);
        expect(screen.getByRole('radio', { checked: true })).toEqual(option);
      });

      it('accepts value as a primitive', () => {
        render(
          <Test
            options={options}
            value="Banana"
            mapOption={({ name }) => ({ label: name })}
          />,
        );
        expect(
          screen.getByRole('radio', { name: 'Banana', checked: true }),
        ).toBeInTheDocument();
      });
    });

    describe('mapOption returns a string', () => {
      const options = [
        { name: 'Apple' },
        { name: 'Banana' },
        { name: 'Orange' },
      ];

      it('maps options', async () => {
        render(
          <Test
            options={options}
            mapOption={({ name }) => name}
          />,
        );
        const option = screen.getByRole('radio', { name: 'Banana' });
        await userEvent.click(option);
        expect(screen.getByRole('radio', { checked: true })).toEqual(option);
      });

      it('accepts value as a primitive', () => {
        render(
          <Test
            options={options}
            value="Banana"
            mapOption={({ name }) => name}
          />,
        );
        expect(
          screen.getByRole('radio', { name: 'Banana', checked: true }),
        ).toBeInTheDocument();
      });
    });
  });
});

describe('required', () => {
  it('sets required', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        required
      />,
    );
    const radio = screen.getByRole('radio');
    expect(radio).toBeRequired();
  });
});

describe('name', () => {
  it('sets the name', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        name="foo"
      />,
    );
    const radio = screen.getByRole('radio');
    expect(radio).toHaveProperty('name', 'foo');
  });
});

describe('onChange', () => {
  it('is called if a radio changes checked state', async () => {
    const options = ['Apple', 'Banana'];
    const onChange = jest.fn();
    render(
      <Test
        options={options}
        onChange={onChange}
        onValue={null}
      />,
    );
    const radio = screen.getByRole('radio', { name: 'Banana' });
    await userEvent.click(radio);
    expect(onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: radio,
      }),
    );
  });
});

describe('placeholderOption', () => {
  it('does not create an option', () => {
    const options = ['Apple', 'Banana'];
    render(
      <Test
        options={options}
        placeholderOption="Foo"
      />,
    );
    expect(
      screen.queryByRole('radio', { name: 'Foo' }),
    ).not.toBeInTheDocument();
  });
});

describe('classPrefix', () => {
  it('when null removes the class', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        classPrefix={null}
      />,
    );
    const radio = screen.getByRole('radio');
    expect(radio).not.toHaveClass();
  });

  it('sets the class', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        classPrefix="bar"
      />,
    );
    const radio = screen.getByRole('radio');
    expect(radio).toHaveClass('bar__input');
  });
});

describe('renderWrapper', () => {
  it('customises the wrapper', () => {
    const spy = jest.fn(({ key, ...props }) => (
      <div
        key={key}
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderWrapper={spy}
        test="foo"
      />,
    );

    expect(screen.getByRole('radio').parentNode).toHaveAttribute(
      'data-foo',
      'bar',
    );
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderInput', () => {
  it('customises the input', () => {
    const spy = jest.fn((props) => (
      <input
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderInput={spy}
        test="foo"
      />,
    );

    expect(screen.getByRole('radio')).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderLabel', () => {
  it('customises the label', () => {
    const spy = jest.fn((props) => (
      <label
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderLabel={spy}
        test="foo"
      />,
    );

    expect(document.querySelector('label')).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderDescription', () => {
  it('customises the description', () => {
    const spy = jest.fn((props) => (
      <div
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', description: 'fizz' }]}
        renderDescription={spy}
        test="foo"
      />,
    );

    const radio = screen.getByRole('radio');
    expect(
      document.getElementById(radio.getAttribute('aria-describedby')),
    ).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderGroup', () => {
  it('customises the group', () => {
    const spy = jest.fn(({ key, ...props }) => (
      <div
        key={key}
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', group: 'fizz' }]}
        renderGroup={spy}
        test="foo"
      />,
    );

    const radio = screen.getByRole('radio');
    expect(radio.parentNode.parentNode).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        group: expect.objectContaining({ label: 'fizz' }),
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderGroupLabel', () => {
  it('customises the group label', () => {
    const spy = jest.fn((props) => (
      <div
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', group: 'fizz' }]}
        renderGroupLabel={spy}
        test="foo"
      />,
    );

    const radio = screen.getByRole('radio');
    expect(radio.parentNode.parentNode.firstElementChild).toHaveAttribute(
      'data-foo',
      'bar',
    );
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        group: expect.objectContaining({ label: 'fizz' }),
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderGroupAccessibleLabel', () => {
  it('customises the group accessible label', () => {
    const spy = jest.fn((props) => (
      <div
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', group: 'fizz' }]}
        renderGroupAccessibleLabel={spy}
        test="foo"
      />,
    );

    expect(document.querySelector('label').firstElementChild).toHaveAttribute(
      'data-foo',
      'bar',
    );
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        group: expect.objectContaining({ label: 'fizz' }),
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});
