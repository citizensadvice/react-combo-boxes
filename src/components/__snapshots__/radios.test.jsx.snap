// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options as array of strings renders radio group 1`] = `
<div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_apple"
      name="name"
      type="radio"
      value="Apple"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_apple"
    >
      Apple
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_banana"
      name="name"
      type="radio"
      value="Banana"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_banana"
    >
      Banana
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_orange"
      name="name"
      type="radio"
      value="Orange"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_orange"
    >
      Orange
    </label>
  </div>
</div>
`;

exports[`options options as array of numbers renders radio group 1`] = `
<div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_0"
      name="name"
      type="radio"
      value="0"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_0"
    >
      0
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_1"
      name="name"
      type="radio"
      value="1"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_1"
    >
      1
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_2"
      name="name"
      type="radio"
      value="2"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_2"
    >
      2
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_3"
      name="name"
      type="radio"
      value="3"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_3"
    >
      3
    </label>
  </div>
</div>
`;

exports[`options options as array of objects group renders grouped options 1`] = `
<div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_apple"
      name="name"
      type="radio"
      value="Apple"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_apple"
    >
      Apple
    </label>
  </div>
  <div
    class="react-combo-boxes-radio-group"
  >
    <div
      class="react-combo-boxes-radio-group__label"
    >
      Citrus
    </div>
    <div
      class="react-combo-boxes-radio"
    >
      <input
        class="react-combo-boxes-radio__input"
        id="id_option_orange"
        name="name"
        type="radio"
        value="Orange"
      />
      <label
        class="react-combo-boxes-radio__label"
        for="id_option_orange"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Citrus 
        </span>
        Orange
      </label>
    </div>
    <div
      class="react-combo-boxes-radio"
    >
      <input
        class="react-combo-boxes-radio__input"
        id="id_option_lemon"
        name="name"
        type="radio"
        value="Lemon"
      />
      <label
        class="react-combo-boxes-radio__label"
        for="id_option_lemon"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Citrus 
        </span>
        Lemon
      </label>
    </div>
  </div>
  <div
    class="react-combo-boxes-radio-group"
  >
    <div
      class="react-combo-boxes-radio-group__label"
    >
      Berry
    </div>
    <div
      class="react-combo-boxes-radio"
    >
      <input
        class="react-combo-boxes-radio__input"
        id="id_option_raspberry"
        name="name"
        type="radio"
        value="Raspberry"
      />
      <label
        class="react-combo-boxes-radio__label"
        for="id_option_raspberry"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Berry 
        </span>
        Raspberry
      </label>
    </div>
    <div
      class="react-combo-boxes-radio"
    >
      <input
        class="react-combo-boxes-radio__input"
        id="id_option_strawberry"
        name="name"
        type="radio"
        value="Strawberry"
      />
      <label
        class="react-combo-boxes-radio__label"
        for="id_option_strawberry"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Berry 
        </span>
        Strawberry
      </label>
    </div>
  </div>
</div>
`;

exports[`options options as array of objects label renders radio group 1`] = `
<div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_apple"
      name="name"
      type="radio"
      value="Apple"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_apple"
    >
      Apple
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_banana"
      name="name"
      type="radio"
      value="Banana"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_banana"
    >
      Banana
    </label>
  </div>
  <div
    class="react-combo-boxes-radio"
  >
    <input
      class="react-combo-boxes-radio__input"
      id="id_option_orange"
      name="name"
      type="radio"
      value="Orange"
    />
    <label
      class="react-combo-boxes-radio__label"
      for="id_option_orange"
    >
      Orange
    </label>
  </div>
</div>
`;
